docs/design/proposals/split-repo/split-repo.md

```diff
@@ -0,0 +1,103 @@
+versioned separately from ACK service controllers and allow individual ACK
+service controllers to implement their own release branching mechanics.
+
+## The Problem
+
+ACK is currently contained in a single monolithic source repository
+(github.com/aws/aws-controllers-k8s). Within this monorepo are a number of
+distinct pieces:
+
+* ACK core API types (`/apis/core`)
+* ACK common controller configuration and runtime (`/pkg/runtime`, `/pkg/config`, `/pkg/types`)
+* ACK code generator (`/cmd/ack-generate`, `/pkg/generate`, `/pkg/model`, `/templates`)
+* ACK common build and test scripts (`/scripts`, `/test`)
+* Individual AWS service-specific controllers (`/services/$SERVICE`)
+* Individual AWS service-specific tests (`/test/e2e/$SERVICE`)
+
+When we generate service controller code in ACK, we build the `ack-generate`
+CLI from the code in this monorepo and output the generated code into the
+`/services/$SERVICE` directory also within this monorepo.
+
+When we build releases in ACK, the releases are Git tags on the monorepo --
+e.g. `v0.0.2` -- and then individual ACK service controller binaries (Docker
+images) are built and pushed to the [ECR public ACK controller image][ecr]
+registry with an image tag of `$SERVICE-$RELEASE_VERSION` -- e.g. `s3-v0.0.2`
+
+[ecr]: https://gallery.ecr.aws/aws-controllers-k8s/controller
+
+### Problem #1: Release series inflexibility
+
+This monorepo tightly couples the ACK common runtime code, code
+generator/templates and the individual service controller code, making the
+release process inflexible and cumbersome. The Elasticache team cannot have
+separate release series -- say, a `stable` and a `preview` release series --
+because releases are Git tags, and Git tags are specific to one source
+repository. If we had a `stable` and `preview` Git branch, all service
+controllers would need to follow the same release cadence and series naming. In
+addition, numeric releases such as `v0.2.0` refer to a specific
+content-addressable SHA1 Git commit, and with all service controllers in the
+same source repository, this means that numeric Git tags cannot applied to two
+different service controllers to represent an incremental change in one
+specific service controller's codebase.
+
+### Problem #2: Common runtime import inflexibility
+
+Furthermore, due to the code generator and ACK common runtime code being in the
+same repo as the service controllers that `import` this code, it's not easy to
+depend on a specific version of that common runtime from within the service
+controller code itself. Individual service controller package directories would
+need their own `go.mod` files and managing this would be a pain and wouldn't
+solve Problem #1 anyway.
+
+## Solution
+
+We propose splitting the monorepo into a set of source code repositories that
+can be separately Git-tagged and have their releases managed according to the
+preferences of the team that owns that component or individual ACK service
```

Comment:
Does this imply that versioning between ACK services could be different? I worry that one team's interpretation of semver and release versioning scheme could differ greatly from another - adding a lot of confusion for devops teams.

Reply:
I agree that keeping something consistent would be best for customers. But if AWS products already have `stable` and `preview` names for their services then they should stick with what customers would know.

Reply:
Chatted with @tabern about this. I think we'd want to establish some guardrails here for service to adhere to. For example, we might say "there will be a `main` Git branch that will contain all commits to the service controller and a `stable` Git branch that will contain commits to the service controller that have 100% e2e test coverage".

Reply:
On the same lines as service controller Git branches, 
`github.com/aws/aws-controllers-k8s-generate` and `github.com/aws/aws-controllers-k8s` should also have `main` and `stable` Git branches. 
This will help service controller's `stable` branch to have its code generated using `stable`  version of `github.com/aws/aws-controllers-k8s-generate` and have dependency on `stable` version of `github.com/aws/aws-controllers-k8s` runtime.

Reply:
https://github.com/aws/aws-controllers-k8s/blob/main/docs/contents/releases.md mentions use of following release statuses:
```
    PROPOSED
    PLANNED
    BUILD
    DEVELOPER PREVIEW
    BETA
    GENERALLY AVAILABLE (GA)
```
The proposed `main`, `stable` branches map to which of the release status?
Should there be branch that correspond to each of these release statuses?

Reply:
> On the same lines as service controller Git branches,
> `github.com/aws/aws-controllers-k8s-generate` and `github.com/aws/aws-controllers-k8s` should also have `main` and `stable` Git branches.

Those are just libraries and will have only a `main` branch with a single SemVer release cadence. Consuming code will reference a specific semver'd release (which is a Git tag) and "pin" their use of the `ack-generate` and common runtime code using standard `go.mod` dependency references.

For example:

https://github.com/aws-controllers-k8s/elasticache-controller/blob/2b2e519d50e2a02f0245d457bf142ed5fa95d537/go.mod#L6

@A-Hilaly is adding the ability to specify a particular aws-sdk-go semver when calling `ack-generate` in order to properly control the specific API definition files that get built for a service controller.

> This will help service controller's `stable` branch to have its code generated using `stable` version of `github.com/aws/aws-controllers-k8s-generate` and have dependency on `stable` version of `github.com/aws/aws-controllers-k8s` runtime.

Totally different things. There is not going to be a `stable` branch of the runtime or the ack-generate tool, just like there isn't a `stable` branch for kubebuilder, k8s.io/controller-runtime or k8s.io/client-go. Instead, code that depends on those libraries import a specific semver release of the library as a dependency in their `go.mod` files.

"stable" branches in the service controller source repos are to allow the service teams to have a branch where they commit all their new feature support to (`main`) and a separate branch that they (potentially more slowly) merge from `main` commits that have been more strenuously tested. The `/helm` and `/config` directories in the service controller repo's "stable" branch will have images that point to earlier Git tags on this stable branch whereas the images in the `main` branch's configuration files will point to the equivalent of a `:latest` tag for that controller.

Reply:
> https://github.com/aws/aws-controllers-k8s/blob/main/docs/contents/releases.md mentions use of following release statuses:
> 
> ```
>     PROPOSED
>     PLANNED
>     BUILD
>     DEVELOPER PREVIEW
>     BETA
>     GENERALLY AVAILABLE (GA)
> ```
> 
> The proposed `main`, `stable` branches map to which of the release status?
> Should there be branch that correspond to each of these release statuses?

I don't think so. I think have a `main` branch for all development and a `stable` branch for code that has "baked" for a bit longer are the only branches that are needed.

Personally, I'd love to move away from the use of the term "Generally Available (GA)" for the vast majority of things and instead primarily use SemVer to communicate API compatibility guarantees and a single term "stable" for conveying our degree of confidence that some bits of code are ready for prolonged and reliable use.

Reply:
Are there guidelines to retire old 'stable' chart/release when new one comes out? Or how many past 'stable' releases are supported?

Reply:
> Are there guidelines to retire old 'stable' chart/release when new one comes out? Or how many past 'stable' releases are supported?

I don't think we'd ever retire semver-numbered Chart release tags -- e.g. `chart/elasticache-v1.2.3` would be a "forever" Helm chart tag. The `chart/elasticache-stable` Helm chart tag would be a single release tag that the deployment.yaml's image would point to different semver-tagged images over time.

Kumar, does that answer your question?

Reply:
As there will be chart that contain both `stable` and the semver, example: `chart/s3-stable-v1.2.6`
My question was about such charts releases.

Reply:
Considering the details on following links:
* https://docs.aws.amazon.com/credref/latest/refdocs/maint-policy.html
* https://docs.aws.amazon.com/credref/latest/refdocs/version-support-matrix.html

My question was to know similar details for ACK Service controllers releases.